File: src/main/java/org/wx/problems/Solution0729.java
package org.wx.problems;

import java.util.TreeMap;

/**
 * 729. 我的日程安排表 I
 */
public class Solution0729 {
    // 区间合并问题，可以利用treeMap解，参考715
    class MyCalendar {
        TreeMap<Integer, Integer> intervals;

        public MyCalendar() {
            intervals = new TreeMap<>();
        }

        public boolean book(int start, int end) {
            if (intervals.isEmpty()) {
                intervals.put(start, end);
                return true;
            }
            if (intervals.containsKey(start)) {
                return false;
            }
            Integer prev = intervals.lowerKey(start);
            Integer next = intervals.higherKey(start);
            if ((prev != null && intervals.get(prev) > start) || (next != null && next < end)) {
                return false;
            }
            intervals.put(start, end);
            return true;
        }
    }

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */
}

File: python/solution_0729.py
class MyCalendar:
    # 直接遍历
    def __init__(self):
        self.booked = []

    def book(self, start: int, end: int) -> bool:
        if any(l < end and start < r for l, r in self.booked):
            return False
        self.booked.append((start, end))
        return True


from sortedcontainers import SortedDict


class MyCalendarV2:
    # 二分查找
    def __init__(self):
        self.booked = SortedDict()

    def book(self, start: int, end: int) -> bool:
        i = self.booked.bisect_left(end)
        if i == 0 or self.booked.items()[i - 1][1] <= start:
            self.booked[start] = end
            return True
        return False


class MyCalendar:
    # 动态线段树
    def __init__(self):
        self.tree = set()
        self.lazy = set()

    def query(self, start: int, end: int, l: int, r: int, idx: int) -> bool:
        if r < start or end < l:
            return False
        if idx in self.lazy:  # 如果该区间已被预订，则直接返回
            return True
        if start <= l and r <= end:
            return idx in self.tree
        mid = (l + r) // 2
        return self.query(start, end, l, mid, 2 * idx) or \
               self.query(start, end, mid + 1, r, 2 * idx + 1)

    def update(self, start: int, end: int, l: int, r: int, idx: int) -> None:
        if r < start or end < l:
            return
        if start <= l and r <= end:
            self.tree.add(idx)
            self.lazy.add(idx)
        else:
            mid = (l + r) // 2
            self.update(start, end, l, mid, 2 * idx)
            self.update(start, end, mid + 1, r, 2 * idx + 1)
            self.tree.add(idx)
            if 2 * idx in self.lazy and 2 * idx + 1 in self.lazy:
                self.lazy.add(idx)

    def book(self, start: int, end: int) -> bool:
        if self.query(start, end - 1, 0, 10 ** 9, 1):
            return False
        self.update(start, end - 1, 0, 10 ** 9, 1)
        return True


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(startTime,endTime)